### PR TITLE
Optimize traits.d by factoring out two common sub-expressions when ev…

### PR DESCRIPTION
**Pull request by nordlow**
…aluating the builtin traits isFinalClass and isTemplate.

### CODE REVIEW

File: compiler/src/dmd/traits.d [hunk 1]
```diff
@@ -482,13 +482,19 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
     }
     if (e.ident == Id.isAbstractClass)
     {
-        return isTypeX(t => t.toBasetype().isTypeClass() &&
-                            t.toBasetype().isTypeClass().sym.isAbstract());
+        return isTypeX((t)
+        {
+            auto c = t.toBasetype().isTypeClass();
```
Comment (thewilsonator) [inline compiler/src/dmd/traits.d:487]: these could be `const`

File: compiler/src/dmd/traits.d [hunk 2]
```diff
@@ -482,13 +482,19 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
...
+            return c && c.sym.isAbstract();
+        });
     }
     if (e.ident == Id.isFinalClass)
     {
-        return isTypeX(t => t.toBasetype().isTypeClass() &&
-                            (t.toBasetype().isTypeClass().sym.storage_class & STC.final_) != 0);
+        return isTypeX((t)
+        {
+            auto c = t.toBasetype().isTypeClass();
```
Comment (thewilsonator) [inline compiler/src/dmd/traits.d:495]: ditto

Comment (nordlow) [reply compiler/src/dmd/traits.d:495]: Done.

File: compiler/src/dmd/traits.d (hunk 1 above)
Comment (nordlow) [reply compiler/src/dmd/traits.d:487]: Making `c` const fails as

```
compiler/src/dmd/traits.d(488,41): Error: mutable method `dmd.dclass.ClassDeclaration.isAbstract` is not callable using a `const` object
```

so skipping this.

**Review by thewilsonator — APPROVED** [review]

**Comment by nordlow** [comment]
What to do with seemingly unrelated build error at https://buildkite.com/dlang/dmd/builds/37787#01918936-daab-4bff-a0d5-f5729c6f16fb/114-140?

**Pull request merged**